### PR TITLE
UDN controller: Initialize namespace-tracker on startup

### PR DESCRIPTION
**Pull request by ormergi**
#### What this PR does and why is it needed

Initialize the namespace-tracker when controller starts
as part of Run().

Populate the namespace-tracker according to the cluster
state.
When the controller actually start it will reconcile CUDN
and NAD objects.

This change enable the controller to cleanup stale NADs
when it starts, recover from a crush and reconciling the
cluster state following changes when controller was down.

#### Which issue(s) this PR fixes

Fixes #

Resolve threads on parent PR https://github.com/ovn-org/ovn-kubernetes/pull/4612#discussion_r1781228553 https://github.com/ovn-org/ovn-kubernetes/pull/4612#discussion_r1781228553

#### Special notes for reviewers

This PR rebased on top of #4612, **only last commit is relevant**.

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go [hunk 1]
```diff
@@ -142,6 +143,48 @@ func (c *Controller) Run() error {
...
+	indexedNADs := map[string]netv1.NetworkAttachmentDefinition{}
+	for _, nad := range nads {
+		if nad != nil {
+			indexedNADs[nad.Namespace+"/"+nad.Name] = *nad.DeepCopy()
+		}
+	}
+
+	for _, cudn := range cudns {
+		cudnCopy := cudn.DeepCopy()
+		for nadKey, nad := range indexedNADs {
```
Comment (npinaeva) [inline go-controller/pkg/clustermanager/userdefinednetwork/controller.go:172]: I think you may not need indexedNADs, and just iterate over `nads` from the `List` call

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go [hunk 2]
```diff
@@ -142,6 +143,48 @@ func (c *Controller) Run() error {
...
+	}
+	indexedNADs := map[string]netv1.NetworkAttachmentDefinition{}
+	for _, nad := range nads {
+		if nad != nil {
+			indexedNADs[nad.Namespace+"/"+nad.Name] = *nad.DeepCopy()
+		}
+	}
+
+	for _, cudn := range cudns {
+		cudnCopy := cudn.DeepCopy()
```
Comment (npinaeva) [inline go-controller/pkg/clustermanager/userdefinednetwork/controller.go:171]: I don't think you need to deepcopy here, as we are not changing cudns in this function?

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go [hunk 3]
```diff
@@ -142,6 +143,48 @@ func (c *Controller) Run() error {
...
+		for nadKey, nad := range indexedNADs {
+			if !metav1.IsControlledBy(&nad, cudnCopy) {
+				continue
+			}
+			if _, exists := c.namespaceTracker[cudnCopy.Name]; !exists {
+				c.namespaceTracker[cudnCopy.Name] = sets.New[string]()
+			}
```
Comment (npinaeva) [inline go-controller/pkg/clustermanager/userdefinednetwork/controller.go:178]: just to be consistent with the logic from cudn reconcile, we can add an empty set for every cudn in the outer loop

File: go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go [hunk 4]
```diff
@@ -697,6 +698,61 @@ var _ = Describe("User Defined Network Controller", func() {
...
+				connectedNsNames := []string{"green", "yellow"}
+				connectedLabel := map[string]string{"test.io": "connected"}
+				for _, nsName := range connectedNsNames {
+					ns := testNamespace(nsName)
+					ns.SetLabels(connectedLabel)
+					testObjs = append(testObjs, ns)
+				}
+				rand.Shuffle(len(testObjs), func(i, j int) {
+					testObjs[i], testObjs[j] = testObjs[j], testObjs[i]
+				})
```
Comment (npinaeva) [inline go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go:720]: interesting, what do we achieve with that?

File: go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go [hunk 5]
```diff
@@ -697,6 +698,61 @@ var _ = Describe("User Defined Network Controller", func() {
...
+				}}
+				testObjs = append(testObjs, cudn)
+
+				c := newTestController(renderNadStub(testClusterUdnNAD(cudn.Name, "")), testObjs...)
+
+				for _, nsName := range append(staleNADsNsNames, connectedNsNames...) {
+					nad := testClusterUdnNAD("test", nsName)
+					nad, err := cs.NetworkAttchDefClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nsName).Create(context.Background(), nad, metav1.CreateOptions{})
+					Expect(err).NotTo(HaveOccurred())
+				}
```
Comment (npinaeva) [inline go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go:734]: why not add these to the testObjs?

**Review by npinaeva — COMMENTED** [review]
nice!

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go (hunk 1 above)
Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go:172]: Having a map enable optimize this loop in a way that it wont process a NAD that already been tracked.
When a NAD is tracked it is removed from the map in line 181.

File: go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go (hunk 4 above)
Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go:720]: Its a left over from testing trying to see order doesnt matter and things work as it should, I will removed it sorry for the noise.

**Comment by ormergi** [comment]
Rebase

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go (hunk 2 above)
Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go]: Done

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go (hunk 3 above)
Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go]: Done

File: go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go (hunk 5 above)
Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go]: Done

**Review by ormergi — COMMENTED** [review]
Changes: resolve

File: go-controller/pkg/clustermanager/userdefinednetwork/controller.go (hunk 1 above)
Comment (npinaeva) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go:173]: oh I completely missed that part, I don't think that it is a good idea to modify the map you are iterating over inside the loop.

Comment (npinaeva) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go:173]: somewhat of a grey area https://groups.google.com/g/golang-nuts/c/eJqwQLONhLs but probably should work

Comment (ormergi) [reply go-controller/pkg/clustermanager/userdefinednetwork/controller.go:173]: Added comment to clearfy when we remove iterated entries.

**Comment by npinaeva** [comment]
Looks good now, let's add it to the https://github.com/ovn-org/ovn-kubernetes/pull/4612

**Comment by ormergi** [comment]
> Looks good now, let's add it to the #4612

Done

**Comment by trozet** [comment]
so should we close this in favor of 4612?

**Comment by ormergi** [comment]
Yep, close in favor of #4612 

/close

**Pull request closed without merging**